# FIT5196 Task 2 in Assessment 2
#### Student Name: Yicheng Zhang
#### Student ID: 27699641

Date: 20/04/2018

Version: 1.0

Environment: Python 3.6.5 and Jupyter notebook

Libraries used:
* pandas (for dataframe, included in Anaconda Python 3.6)  
* numpy (for numpy array, included in Anaconda Python 3.6)
* sys (for system information, included in Anaconda Python 3.6)
* datetime(for parse date, included in Anaconda Python 3.6)

## 1.  Import libraries 

In [1]:
import pandas as pd
import numpy as np
import datetime
import sys

In [2]:
sys.version

'3.6.5 |Anaconda custom (64-bit)| (default, Mar 29 2018, 13:32:41) [MSC v.1900 64 bit (AMD64)]'

## 2. Load Data with Pandas

### 2.1. Load CSV File
First step of this assessment is loading the csv file's data from the file, In python,one of the optimized method is using the `pandas` build-in method named `read_csv()`
We can get the data with `dataframe` type.

In [3]:
dataset1 = pd.read_csv('dataset1_solution.csv')
dataset1.head()

Id                                              Title   Location  \
0  12612628                        Engineering Systems Analyst    Dorking   
1  12612830                            Stress Engineer Glasgow    Glasgow   
2  12612844                   Modelling and simulation analyst  Hampshire   
3  12613049  Engineering Systems Analyst / Mathematical Mod...     Surrey   
4  12613647         Pioneer, Miser Engineering Systems Analyst     Surrey   

    ContractType ContractTime                       Company          Category  \
0  non_specified    permanent  Gregory Martin International  Engineering Jobs   
1  non_specified    permanent  Gregory Martin International  Engineering Jobs   
2  non_specified    permanent  Gregory Martin International  Engineering Jobs   
3  non_specified    permanent  Gregory Martin International  Engineering Jobs   
4  non_specified    permanent  Gregory Martin International  Engineering Jobs   

  Salary per annum        SourceName         OpenDate        CloseDate  
0            24996  cv-library.co.uk  20121103T000000  20121203T000000  
1            30000  cv-library.co.uk  20130108T150000  20130408T150000  
2            30000  cv-library.co.uk  20130726T150000  20130924T150000  
3            27504  cv-library.co.uk  20121214T000000  20130314T000000  
4            24996  cv-library.co.uk  20131025T000000  20131224T000000

In [4]:
dataset2 = pd.read_csv('dataset2_integration.csv')
dataset2.head()

Id          Source Name  \
0  69247680  jobs.guardian.co.uk   
1  69247682        icaewjobs.com   
2  69247685  jobs.guardian.co.uk   
3  69247688  jobs.guardian.co.uk   
4  69247694  jobs.guardian.co.uk   

                                               Title        location  \
0  Business Development Exec  research / insight ...          London   
1                               Audit Senior  London          London   
2  PR & Social Media Account Executive  Top Inter...  Central London   
3                                    Content Manager          London   
4                              AV PRODUCTION MANAGER  Worcestershire   

  Contract Type Contract Time               Company  \
0            ft         perm.     BOYCE RECRUITMENT   
1            ft        contr.           Pro Finance   
2            ft         perm.  ECOM RECRUITMENT LTD   
3            ft         perm.         NAKAMA LONDON   
4            ft         perm.      LIVE RECRUITMENT   

                           Category  Salary per month             OpenDate  \
0  PR, Advertising & Marketing Jobs              2125  2012-03-01 15:00:00   
1         Finance & Accounting Jobs              3600  2013-03-14 12:00:00   
2  PR, Advertising & Marketing Jobs              1917  2012-02-14 12:00:00   
3  PR, Advertising & Marketing Jobs              2083  2012-10-21 00:00:00   
4  PR, Advertising & Marketing Jobs              2292  2012-02-22 00:00:00   

             CloseDate  
0  2012-03-31 15:00:00  
1  2013-04-13 12:00:00  
2  2012-03-15 12:00:00  
3  2012-12-20 00:00:00  
4  2012-04-22 00:00:00

In [5]:
print (dataset1.shape)

(25077, 11)


In [6]:
print (dataset2.shape)

(25276, 11)


There are total 25077 records in dataset1, and 11 columns.

There are total 25276 records in dataset2, and 11 columns. 

### 2.2. Check Data information
There are many methods to check the data information, they can be statistic, or descriptive. The `pandas` package provides us a convenient way to observe data information easily.
- - -
`dataframe.info()` can show `The number of records`, `Columns(Name,Type,Non-null values)`,`memory usage`

The dataset1 we have checked before, there are some differences between two datasets. Such as, the type of `Salary per month` and type of `Salary per annum`; There are many null value in dataset2 in `Contract Type` and `Contract Time` columns

In [7]:
dataset1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25077 entries, 0 to 25076
Data columns (total 11 columns):
Id                  25077 non-null int64
Title               25077 non-null object
Location            25077 non-null object
ContractType        25077 non-null object
ContractTime        25077 non-null object
Company             21242 non-null object
Category            25077 non-null object
Salary per annum    25077 non-null object
SourceName          25077 non-null object
OpenDate            25077 non-null object
CloseDate           25077 non-null object
dtypes: int64(1), object(10)
memory usage: 2.1+ MB


In [8]:
dataset2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25276 entries, 0 to 25275
Data columns (total 11 columns):
Id                  25276 non-null int64
Source Name         25276 non-null object
Title               25276 non-null object
location            25276 non-null object
Contract Type       4819 non-null object
Contract Time       19763 non-null object
Company             24178 non-null object
Category            25276 non-null object
Salary per month    25276 non-null int64
OpenDate            25276 non-null object
CloseDate           25276 non-null object
dtypes: int64(2), object(9)
memory usage: 2.1+ MB


In [9]:
dataset1.describe()

Id
count  2.507700e+04
mean   6.664312e+07
std    5.195261e+06
min    1.261263e+07
25%    6.720830e+07
50%    6.836110e+07
75%    6.871371e+07
max    6.924767e+07

In [10]:
dataset2.describe()

Id  Salary per month
count  2.527600e+04      25276.000000
mean   7.109038e+07       2935.074458
std    1.025040e+06       1315.875990
min    6.924768e+07        417.000000
25%    7.008671e+07       2000.000000
50%    7.129211e+07       2681.500000
75%    7.187785e+07       3542.000000
max    7.270520e+07       8000.000000

The Id column in dataset2 seems to be after the dataset1, there is no duplicate Id according to the above describe.

**Discovering a further discovery using descibe objects.**

In [11]:
dataset1.describe(include='O')

Title Location   ContractType ContractTime  \
count                   25077    25077          25077        25077   
unique                  25077      477              3            3   
top     Semi Senior  TJ****SN       UK  non_specified    permanent   
freq                        1     3996          19499        16194   

              Company Category Salary per annum     SourceName  \
count           21242    25077            25077          25077   
unique           4879        8             1574             89   
top     UKStaffsearch  IT Jobs            35004  totaljobs.com   
freq              248     7085             1011           5335   

               OpenDate        CloseDate  
count             25077            25077  
unique             2193             2400  
top     20130127T150000  20130526T120000  
freq                 27               23

In [12]:
dataset2.describe(include='O')

Source Name                                 Title location  \
count           25276                                 25276    25276   
unique             92                                 25276      475   
top     totaljobs.com  White Goods Technical Support Export       UK   
freq             4799                                     1     4263   

       Contract Type Contract Time        Company Category  \
count           4819         19763          24178    25276   
unique             2             2           5679        8   
top               ft         perm.  UKStaffsearch  IT Jobs   
freq            4353         16505            255     6854   

                   OpenDate            CloseDate  
count                 25276                25276  
unique                 2193                 2395  
top     2012-04-06 12:00:00  2013-07-31 12:00:00  
freq                     28                   25

It can be found that there are slightly differences between two datasets.
* The column's oreder is different.
* The format of Date is different.
* The format of Contract type/time is different.

**Check the columns' name**

In [13]:
dataset1.columns.tolist()

['Id',
 'Title',
 'Location',
 'ContractType',
 'ContractTime',
 'Company',
 'Category',
 'Salary per annum',
 'SourceName',
 'OpenDate',
 'CloseDate']

In [14]:
dataset2.columns.tolist()

['Id',
 'Source Name',
 'Title',
 'location',
 'Contract Type',
 'Contract Time',
 'Company',
 'Category',
 'Salary per month',
 'OpenDate',
 'CloseDate']

Checking the Column's name, we can find that there are some differences:

Assume we should fit dataset2 to dataset1, we can map columns semantically as: 
* **`'location'` -> `'Location'`**
* **`'Contract Type'` -> `'ContractType'`**
* **`'Contract Time'` -> `'ContractTime'`**
* **`'Source Name'` -> `'SourceName'`**
- - -
`Salary per month` in the dataset 2 may be aggregateed and caculated to fir dataset1's `Salary per annum`

## 3. Resolving schema conflicts and merging data

### 3.1. Resolving schema conflicts

**From the above checking, we can find some schema conflicts, we should fix the conflicts column by column**

**First, fix data type/namespace conflicts**

In [15]:
print(dataset1.ContractType.value_counts())
print(dataset2['Contract Type'].value_counts())

non_specified    19499
full_time         4883
part_time          695
Name: ContractType, dtype: int64
ft    4353
pt     466
Name: Contract Type, dtype: int64


In [16]:
print(dataset1.ContractTime.value_counts())
print(dataset2['Contract Time'].value_counts())

permanent        16194
non_specified     6212
contract          2671
Name: ContractTime, dtype: int64
perm.     16505
contr.     3258
Name: Contract Time, dtype: int64


In [17]:
print(dataset1.Category.value_counts())
print(dataset2.Category.value_counts())

IT Jobs                             7085
Healthcare & Nursing Jobs           4334
Engineering Jobs                    3458
Accounting & Finance Jobs           3099
Sales Jobs                          2609
Hospitality & Catering Jobs         2124
Teaching Jobs                       1378
PR, Advertising & Marketing Jobs     990
Name: Category, dtype: int64
IT Jobs                             6854
Engineering Jobs                    4199
Healthcare & Nursing Jobs           3984
Finance & Accounting Jobs           2842
Hospitality & Catering Jobs         2310
Sales Jobs                          2268
Teaching Jobs                       1890
PR, Advertising & Marketing Jobs     929
Name: Category, dtype: int64


There are conflicts in some of the columns:
* **Contract Type** 

null values replace by `non_specified`, format(ft -> full_time,pt -> part_time).
- - -
* **Contract Time** 

null values replace by `non_specified`, format(perm. -> permanent,contr. -> contract)
- - -
* **Category**

'Accounting & Finance Jobs' -> 'Finance & Accounting Jobs'
- - -
* **Salary per month**

It should be cauclated as salary per anniversary. Guess the relationship between salary per month and salary per anniversary should be 12 times(12 month).
- - -
* **Date**

The date format in dataset2 is `2012-04-06 12:00:00`, we should transfer this format to dataset1's format:`20120406T120000`

**Change the contract columns**

In [18]:
replacetype = {'ft' : 'full_time','pt' : 'part_time'}
dataset2['Contract Type'].replace(replacetype,inplace = True)
dataset2['Contract Type'].fillna('non_specified',inplace=True)

replacetime = {'perm.' : 'permanent','contr.' : 'contract'}
dataset2['Contract Time'].replace(replacetime,inplace = True)
dataset2['Contract Time'].fillna('non_specified',inplace=True)

In [19]:
dataset2['Contract Type'].value_counts()

non_specified    20457
full_time         4353
part_time          466
Name: Contract Type, dtype: int64

In [20]:
dataset2['Contract Time'].value_counts()

permanent        16505
non_specified     5513
contract          3258
Name: Contract Time, dtype: int64

**Change category**

In [21]:
dataset2.Category.replace({'Finance & Accounting Jobs': 'Accounting & Finance Jobs'},inplace=True)
dataset2.Category.value_counts()

IT Jobs                             6854
Engineering Jobs                    4199
Healthcare & Nursing Jobs           3984
Accounting & Finance Jobs           2842
Hospitality & Catering Jobs         2310
Sales Jobs                          2268
Teaching Jobs                       1890
PR, Advertising & Marketing Jobs     929
Name: Category, dtype: int64

**Change Salary**

Adding a new column to store the `Salary per annum`

In [22]:
dataset2['Salary per annum'] = dataset2['Salary per month'].apply(lambda x: x * 12) #salary per annual 

**Change Date**

create another two columns to store date format.

In [23]:
dataset2['OpenDate_parsed'] = pd.to_datetime(dataset2.OpenDate,format="%Y-%m-%d %H:%M:%S",errors='coerce')
dataset2['CloseDate_parsed'] = pd.to_datetime(dataset2.CloseDate,format="%Y-%m-%d %H:%M:%S",errors='coerce')

Replace date by using strftime.

In [24]:
dataset2['OpenDate'] = dataset2['OpenDate_parsed'].apply(lambda x: x.strftime("%Y%m%dT%H%M%S"))

In [25]:
dataset2['CloseDate'] = dataset2['CloseDate_parsed'].apply(lambda x: x.strftime("%Y%m%dT%H%M%S"))

In [26]:
dataset2.head()

Id          Source Name  \
0  69247680  jobs.guardian.co.uk   
1  69247682        icaewjobs.com   
2  69247685  jobs.guardian.co.uk   
3  69247688  jobs.guardian.co.uk   
4  69247694  jobs.guardian.co.uk   

                                               Title        location  \
0  Business Development Exec  research / insight ...          London   
1                               Audit Senior  London          London   
2  PR & Social Media Account Executive  Top Inter...  Central London   
3                                    Content Manager          London   
4                              AV PRODUCTION MANAGER  Worcestershire   

  Contract Type Contract Time               Company  \
0     full_time     permanent     BOYCE RECRUITMENT   
1     full_time      contract           Pro Finance   
2     full_time     permanent  ECOM RECRUITMENT LTD   
3     full_time     permanent         NAKAMA LONDON   
4     full_time     permanent      LIVE RECRUITMENT   

                           Category  Salary per month         OpenDate  \
0  PR, Advertising & Marketing Jobs              2125  20120301T150000   
1         Accounting & Finance Jobs              3600  20130314T120000   
2  PR, Advertising & Marketing Jobs              1917  20120214T120000   
3  PR, Advertising & Marketing Jobs              2083  20121021T000000   
4  PR, Advertising & Marketing Jobs              2292  20120222T000000   

         CloseDate  Salary per annum     OpenDate_parsed    CloseDate_parsed  
0  20120331T150000             25500 2012-03-01 15:00:00 2012-03-31 15:00:00  
1  20130413T120000             43200 2013-03-14 12:00:00 2013-04-13 12:00:00  
2  20120315T120000             23004 2012-02-14 12:00:00 2012-03-15 12:00:00  
3  20121220T000000             24996 2012-10-21 00:00:00 2012-12-20 00:00:00  
4  20120422T000000             27504 2012-02-22 00:00:00 2012-04-22 00:00:00

Drop the cloumns we crated before, and column we do not need.

In [27]:
dataset2.drop(columns=['OpenDate_parsed', 'CloseDate_parsed','Salary per month'],inplace=True)

In [28]:
dataset2.head()

Id          Source Name  \
0  69247680  jobs.guardian.co.uk   
1  69247682        icaewjobs.com   
2  69247685  jobs.guardian.co.uk   
3  69247688  jobs.guardian.co.uk   
4  69247694  jobs.guardian.co.uk   

                                               Title        location  \
0  Business Development Exec  research / insight ...          London   
1                               Audit Senior  London          London   
2  PR & Social Media Account Executive  Top Inter...  Central London   
3                                    Content Manager          London   
4                              AV PRODUCTION MANAGER  Worcestershire   

  Contract Type Contract Time               Company  \
0     full_time     permanent     BOYCE RECRUITMENT   
1     full_time      contract           Pro Finance   
2     full_time     permanent  ECOM RECRUITMENT LTD   
3     full_time     permanent         NAKAMA LONDON   
4     full_time     permanent      LIVE RECRUITMENT   

                           Category         OpenDate        CloseDate  \
0  PR, Advertising & Marketing Jobs  20120301T150000  20120331T150000   
1         Accounting & Finance Jobs  20130314T120000  20130413T120000   
2  PR, Advertising & Marketing Jobs  20120214T120000  20120315T120000   
3  PR, Advertising & Marketing Jobs  20121021T000000  20121220T000000   
4  PR, Advertising & Marketing Jobs  20120222T000000  20120422T000000   

   Salary per annum  
0             25500  
1             43200  
2             23004  
3             24996  
4             27504

Rename the columns with different names.

In [29]:
dataset2 = dataset2.rename(columns = 
                 {'Source Name':'SourceName',
                  'Contract Type':'ContractType',
                  'Contract Time':'ContractTime',
                  'location':'Location'})

In [30]:
dataset2.head()

Id           SourceName  \
0  69247680  jobs.guardian.co.uk   
1  69247682        icaewjobs.com   
2  69247685  jobs.guardian.co.uk   
3  69247688  jobs.guardian.co.uk   
4  69247694  jobs.guardian.co.uk   

                                               Title        Location  \
0  Business Development Exec  research / insight ...          London   
1                               Audit Senior  London          London   
2  PR & Social Media Account Executive  Top Inter...  Central London   
3                                    Content Manager          London   
4                              AV PRODUCTION MANAGER  Worcestershire   

  ContractType ContractTime               Company  \
0    full_time    permanent     BOYCE RECRUITMENT   
1    full_time     contract           Pro Finance   
2    full_time    permanent  ECOM RECRUITMENT LTD   
3    full_time    permanent         NAKAMA LONDON   
4    full_time    permanent      LIVE RECRUITMENT   

                           Category         OpenDate        CloseDate  \
0  PR, Advertising & Marketing Jobs  20120301T150000  20120331T150000   
1         Accounting & Finance Jobs  20130314T120000  20130413T120000   
2  PR, Advertising & Marketing Jobs  20120214T120000  20120315T120000   
3  PR, Advertising & Marketing Jobs  20121021T000000  20121220T000000   
4  PR, Advertising & Marketing Jobs  20120222T000000  20120422T000000   

   Salary per annum  
0             25500  
1             43200  
2             23004  
3             24996  
4             27504

In dataset1, we store the salary as string(contain range salary), we should also modify the type of salary in dataset2 as the string(or object).

In [31]:
dataset2['Salary per annum'] = dataset2['Salary per annum'].astype(str) #fit data type with dataset1

In [32]:
dataset2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25276 entries, 0 to 25275
Data columns (total 11 columns):
Id                  25276 non-null int64
SourceName          25276 non-null object
Title               25276 non-null object
Location            25276 non-null object
ContractType        25276 non-null object
ContractTime        25276 non-null object
Company             24178 non-null object
Category            25276 non-null object
OpenDate            25276 non-null object
CloseDate           25276 non-null object
Salary per annum    25276 non-null object
dtypes: int64(1), object(10)
memory usage: 2.1+ MB


Try to merge two tables using outer join. on dataset1's column.

In [33]:
mergedf = pd.merge(dataset1,dataset2,how='outer',on=dataset1.columns.tolist())

Check the dataset after we merge them togther.

In [34]:
mergedf.describe(include='O')

Title Location   ContractType  \
count                                        50353    50353          50353   
unique                                       50153      477              3   
top     Manufacturing Engineer (Heavy Engineering)       UK  non_specified   
freq                                             2     8259          39956   

       ContractTime        Company Category Salary per annum     SourceName  \
count         50353          45420    50353            50353          50353   
unique            3           7483        8             2231             95   
top       permanent  UKStaffsearch  IT Jobs            35004  totaljobs.com   
freq          32699            503    13939             1930          10134   

               OpenDate        CloseDate  
count             50353            50353  
unique             2193             2413  
top     20130127T150000  20120908T150000  
freq                 41               40

From the above descibe function, we can found that there are some records have same `Title`, there may be possiblity that there are duplicate records in two datasets.
- - -
To identify which rows are duplicated, we should select a global key for the integrated data to decide what attribute make two records the duplicated record.
- - -
`Title` is one of the main attribute can identify record, however, there may be some different companies require similar job by using same title. We should use other attributes to restrict record. 
- - - 
Considering about the real world job advertisement, if we have a job in the same location, same company, with same category and salary, we can have high chance to identify unique job position.
- - -
In this case, we can add `Loacation`,`Category`,`Salary per annum`,`Company` as global key to identify duplicated records.

In [35]:
mergedf[mergedf.duplicated(['Title','Location','Category','Salary per annum','Company'],keep = 'first')] #mark duplicates excpet first occur

Id                                              Title  \
25169  69249686                              Trading Administrator   
25210  69249874                          Staff Nurse  BexhillonSea   
25243  69250133  Support Care Workers, Luton Bedford (Evenings ...   
25896  69534632  Java Software Engineer  Java, Web Services, J*...   
26121  69540424  Perm RGN s Huddersfield **** ph Excellent Nati...   
26149  69541199                                    Staff Nurse RMN   
26161  69541267           TRAINEE IT MANAGER  SYSTEMS ADMIN  ****k   
26211  69544173                                     Senior Chemist   
26242  69544594  NVQ Assessor Banking/Financial Services Salary...   
26295  69545085  Assistant Manager ST ANDREWS, FIFE Restaurant ...   
26518  69556650  Permenant Recruitment Consultant  Commercial D...   
26538  69556933  NHS LSMS Specialist (Local Security Management...   
26549  69557174                        Powertrain Project Engineer   
26605  69559108  Year **** Class Teacher Urgently Required for ...   
26748  69570901                                     German Speaker   
27148  69593348                 ConsultantLondon Markets Insurance   
27220  69596435                   Maintenance Engineer  Electrical   
27223  69596503  Project Manager / Software Implementation Cons...   
27455  69614996          Clinical Nurse Trainer Renal Dialysis/ITU   
27589  69621828  Assistant Manager  Carluccio's  Aberdeen  New ...   
27710  69638838               Eon  Venue & Events Sales Consultant   
27716  69639018                           Wealth Financial Advisor   
27751  69640929       French and German Speaking Credit Controller   
28139  69674149          Staff Nurse  specialist care  Southampton   
28246  69684016  Service Engineer (White Goods, Electric/Gas Ap...   
28267  69685781          Mechanical Engineer / Project Coordinator   
28335  69687111  New Senior Software Developer (C / ASPNET) Nee...   
28564  69693243               Minehead, Somerset  Pharmacy Manager   
28668  69731878     Science Technician / Practical Science Teacher   
28806  69743315                   NET Web Developer ASPNET / VBNET   
...         ...                                                ...   
44748  72117284              HEAD CHEF  PREMIUM BRANDED RESTAURANT   
44818  72118451  DEPUTY PREMISES OFFICER FOR HISTORIC OFFICES  ...   
44952  72127236   IT Applications Supplier Manager, Vendor Manager   
45015  72127656                         Sales Engineer  Pneumatics   
45130  72128412  Chefs Required Immediate Start Heathrow Area ****   
45182  72128669                     FPGA Engineer  Gloucestershire   
45350  72161309  Assistant ManagerNew Restaurant Opening Isling...   
45480  72184151                      Delivery Project Manager/Lead   
45542  72186969   Senior Electronics Design & Development Engineer   
45848  72231674                        Systems Definition Engineer   
46313  72243182   NET Developer  Prestigious Media Brand  Solihull   
46515  72248745       OO Perl Developer  New Start up London ****k   
46542  72249246  Demi Chef de Partie  High Quality Brasserie Di...   
46656  72325981  Demi Chef de Partie  Luxury country House Hote...   
46673  72327631  Production Print Device Sales Manager / Field ...   
46819  72344120                               Community Care Staff   
46950  72373010                    Sales Engineer  Instrumentation   
47110  72394613        C / ASPNET Developer,Manchester, ****  ****   
47219  72400486  Junior Sales and Customer Service Roles  No Ex...   
47330  72406257  Software Developer  C, ASPNet  Permanent  Glas...   
47362  72426076                  Nursery Teacher With RML Training   
47385  72428167  Healthcare Assistants  Care & Nursing Homes  C...   
47577  72439845             Vehicle Technician / Mechanic  Norwich   
47744  72441982                             Account Manager  South   
47783  72442334  Infrastructure & Unix Tech Support (Solaris, U...   
48139  72446581        

Pick up one of the record as an example to verify our assumption

In [36]:
mergedf[mergedf.Title == 'RGN/RMN']

Id    Title Location   ContractType   ContractTime Company  \
1078   51285910  RGN/RMN   Bolton  non_specified  non_specified     NaN   
29742  69849495  RGN/RMN   Bolton  non_specified  non_specified     NaN   

                        Category Salary per annum      SourceName  \
1078   Healthcare & Nursing Jobs            14676  careworx.co.uk   
29742  Healthcare & Nursing Jobs            14676  careworx.co.uk   

              OpenDate        CloseDate  
1078   20130911T150000  20130925T150000  
29742  20130911T150000  20130925T150000

Drop the diplicates records.

In [37]:
mergedf.drop_duplicates(['Title','Location','Category','Salary per annum','Company'],inplace=True) # drop duplicate records

In [38]:
mergedf.reset_index(inplace=True,drop=True) #reset the index.

In [39]:
mergedf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50153 entries, 0 to 50152
Data columns (total 11 columns):
Id                  50153 non-null int64
Title               50153 non-null object
Location            50153 non-null object
ContractType        50153 non-null object
ContractTime        50153 non-null object
Company             45254 non-null object
Category            50153 non-null object
Salary per annum    50153 non-null object
SourceName          50153 non-null object
OpenDate            50153 non-null object
CloseDate           50153 non-null object
dtypes: int64(1), object(10)
memory usage: 4.2+ MB


## 4. Save the data

Save the data to csv file and using the default encoding

In [40]:
mergedf.to_csv('dataset1_dataset2_solution.csv',encoding='utf-8',index=False)

## 5. References

[pandas](http://pandas.pydata.org/pandas-docs/stable/tutorials.html)
- - -
[Reading Materials from moodle](https://moodle.vle.monash.edu/course/view.php?id=42906&section=13#13)
- - -
[numpy](https://docs.scipy.org/doc/numpy/user/quickstart.html)
- - -
[python doc.](https://docs.python.org/3/library/index.html)